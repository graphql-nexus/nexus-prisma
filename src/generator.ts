import { DMMF } from '@prisma/client/runtime'
import endent from 'endent'
import * as fs from 'fs-jetpack'
import * as OS from 'os'
import * as Path from 'path'
import * as tsm from 'ts-morph'
import { d } from './helpers/debugNexusPrisma'
import { SettingsData } from './settings'

const OUTPUT_SOURCE_DIR = Path.join(__dirname, 'runtime')

/**
 * Generate the Nexus Prisma runtime.
 *
 * This will transform the given DMMF into JS source code with accompanying TS declarations. This will be
 * emitted into a "hole" in the internal package source tree.
 */
export function generateRuntime(dmmf: DMMF.Document, _settings: SettingsData): void {
  d('start generateRuntime')

  const generatedSource = generateRuntimeSource(dmmf)

  d('done generateRuntime')

  d('start create TS project and emit transpiled JS')

  const tsProject = new tsm.Project({
    compilerOptions: {
      declaration: true,
      declarationMap: true,
    },
    useInMemoryFileSystem: true,
  })

  tsProject.createSourceFile(
    'index.ts',
    endent`
      export * as models from './models'
    `
  )

  tsProject.createSourceFile('models.ts', generatedSource)

  const emitResult = tsProject.emitToMemory()

  d('done create TS project and emit transpiled JS')

  d(`start writing emitted files to ${OUTPUT_SOURCE_DIR}`)

  emitResult.getFiles().forEach((emittedFile) => {
    const packageModulePath = Path.join(OUTPUT_SOURCE_DIR, emittedFile.filePath)
    fs.remove(packageModulePath)
    fs.write(packageModulePath, emittedFile.text)
    d(`done writing emitted file ${packageModulePath}`)
  })

  d(`done writing all emitting files`)
}

/** Transform the given DMMF into TS source code. */
export function generateRuntimeSource(dmmf: DMMF.Document): string {
  return dmmf.datamodel.models.map(modelToTypeScriptSource).join(OS.EOL + OS.EOL)
}

/** Transform a DMMF model into TS source code. */
function modelToTypeScriptSource(model: DMMF.Model): string {
  return endent`
    export const ${model.name} = {
      ${model.fields.map(fieldToTypeScriptSource).join(`,${OS.EOL}`)},
    } as const
  `
}

/**
 * Transform a DMMF model _field_ into TS source code.
 *
 * @example
 *   // Example of field data given:
 *   const field = {
 *     name: 'f16',
 *     kind: 'scalar',
 *     isList: false,
 *     isRequired: true,
 *     isUnique: false,
 *     isId: false,
 *     isReadOnly: false,
 *     type: 'DateTime',
 *     hasDefaultValue: false,
 *     isGenerated: false,
 *     isUpdatedAt: false,
 *   }
 */
function fieldToTypeScriptSource(field: DMMF.Field) {
  return endent`
    ${field.name}: {
      name: '${field.name}',
      type: '${field.type}',
      isList: ${field.isList},
      isRequired: ${field.isRequired},
      isId: ${field.isId},
      isUnique: ${field.isUnique},
      hasDefaultValue: ${field.hasDefaultValue},
      documentation: ${field.documentation ?? null},
    }
  `
}
