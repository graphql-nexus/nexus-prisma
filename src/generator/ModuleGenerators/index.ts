export * as ModuleGenerators from './index_'
