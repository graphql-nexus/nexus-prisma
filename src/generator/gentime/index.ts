export * as Gentime from './index_'
