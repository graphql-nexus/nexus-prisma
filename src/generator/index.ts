export { generateRuntimeAndEmit } from './generate'
export { settings } from './settingsSingleton'
