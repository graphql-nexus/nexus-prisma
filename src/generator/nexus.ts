import { DMMF } from '@prisma/client/runtime'
import endent from 'endent'
import * as fs from 'fs-jetpack'
import * as OS from 'os'
import { GeneratedModule } from './types'

/** Transform the given DMMF into TS source code. */
export function generateRuntimeNexusSource(dmmf: DMMF.Document): GeneratedModule {
  if (process.env.NP_DEBUG) {
    fs.write('dmmf.json', dmmf)
  }

  const content = endent`
    import * as Nexus from 'nexus'

    ${dmmf.datamodel.models.map(modelToTypeScriptSource).join(OS.EOL + OS.EOL)}
  `

  return {
    fileName: 'nexus.ts',
    content,
  }
}

/** Transform a DMMF model into TS source code. */
function modelToTypeScriptSource(model: DMMF.Model): string {
  return endent`

    export const ${model.name} = {
      ${model.fields.filter(isSupportedType).map(fieldToTypeScriptSource).join(`,${OS.EOL}`)},
    } as const
  `
}

/**
 * Transform a DMMF model _field_ into TS source code.
 *
 * @example
 *   // Example of field data given:
 *   const field = {
 *     name: 'f16',
 *     kind: 'scalar',
 *     isList: false,
 *     isRequired: true,
 *     isUnique: false,
 *     isId: false,
 *     isReadOnly: false,
 *     type: 'DateTime',
 *     hasDefaultValue: false,
 *     isGenerated: false,
 *     isUpdatedAt: false,
 *   }
 */
function fieldToTypeScriptSource(field: DMMF.Field) {
  return endent`
    ${field.name}: {
      name: '${field.name}',
      config: {
        type: ${renderType(field)},
      }
    }
  `
}

function renderType(field: DMMF.Field): string {
  const graphqlType = getGraphQLType(field)

  if (field.isList) {
    return endent`
      Nexus.nonNull(Nexus.list(Nexus.nonNull(${quote(graphqlType)})))
    `
  } else if (field.isRequired) {
    return endent`
      Nexus.nonNull(${quote(graphqlType)})
    `
  } else {
    return endent`
      Nexus.nullable(${quote(graphqlType)})
    `
  }
}

type PrismaScalarType =
  | 'String'
  | 'Boolean'
  | 'Int'
  | 'BigInt'
  | 'Float'
  | 'Decimal'
  | 'DateTime'
  | 'Json'
  | 'Bytes'

function quote(x: string): string {
  return `'${x}'`
}

type GraphQLScalarType = 'ID' | 'String' | 'Int' | 'Float' | 'Boolean'

function isSupportedType(field: DMMF.Field): boolean {
  return field.kind === 'scalar'
}

function getGraphQLType(field: DMMF.Field): GraphQLScalarType {
  if (field.kind !== 'scalar') {
    throw new Error(`type mapping only supported for scalar-type fields`)
  }

  const typeName = field.type as PrismaScalarType

  switch (typeName) {
    case 'String': {
      if (field.isId) {
        return GraphQLScalarTypes.ID
      }
      return GraphQLScalarTypes.String
    }
    default: {
      return GraphQLScalarTypes.String
    }
  }
}

const GraphQLScalarTypes = {
  ID: 'ID',
  String: 'String',
} as const
