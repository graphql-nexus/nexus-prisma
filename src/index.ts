import { enforceValidPeerDependencies } from './lib/peerDepValidator'
import * as nexusPrisma from './nexusPrisma'

enforceValidPeerDependencies({
  packageJson: require('../package.json'),
})

export { nexusPrisma }

export default nexusPrisma
