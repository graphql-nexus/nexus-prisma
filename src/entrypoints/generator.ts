import { Gentime } from '../generator/gentime/settingsSingleton'

export const settings = Gentime.changeSettings
