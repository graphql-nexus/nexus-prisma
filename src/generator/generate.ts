import * as fs from 'fs-jetpack'
import * as Path from 'path'
import * as pkgup from 'pkg-up'

import { DMMF } from '@prisma/client/runtime'

import { d } from '../helpers/debugNexusPrisma'
import { Gentime } from './gentime'
import { ModuleGenerators } from './ModuleGenerators'
import { Module } from './types'

const OUTPUT_SOURCE_DIR_ESM = getOutputSourceDir({ esm: true })
const OUTPUT_SOURCE_DIR_CJS = getOutputSourceDir({ esm: false })

/**
 * Find the output source directory. When using the Yalc workflow some additional hacking around is required.
 *
 * If the Yalc issue https://github.com/wclr/yalc/issues/156 is resolved then this should be as simple as
 * using __dirname.
 */
function getOutputSourceDir(params: { esm: boolean }): string {
  let outputSourceDir: string

  if (process.env.npm_package_dependencies_nexus_prisma === 'file:.yalc/nexus-prisma') {
    const packageJsonFilePath = pkgup.sync()
    if (packageJsonFilePath === null) {
      throw new Error(
        `Nexus Prisma error: Could not find the project root. Project root is the nearest ancestor directory to where this module is running (${__filename}) containing a package.json. Without this information Nexus Prisma does not know where to output its generated code.`
      )
    }
    outputSourceDir = Path.join(
      Path.dirname(packageJsonFilePath),
      params.esm ? 'node_modules/nexus-prisma/dist-esm/runtime' : 'node_modules/nexus-prisma/dist-cjs/runtime'
    )
  } else {
    /**
     * At this point in the code we don't know if the CLI running is the CJS or ESM version.
     * If it is the CJS version and we're doing an ESM build then we need to adjust the __dirname value.
     */
    outputSourceDir = Path.join(
      params.esm ? __dirname.replace('dist-cjs', 'dist-esm') : __dirname,
      '../runtime'
    )
  }

  d(`found outputSourceDir ${outputSourceDir}`)

  return outputSourceDir
}

/** Generate the Nexus Prisma runtime files and emit them into a "hole" in the internal package source tree. */
export function generateRuntimeAndEmit(dmmf: DMMF.Document, settings: Gentime.Settings.Manager): void {
  d('start generateRuntime with configuration %j', settings)

  d('start generateRuntime')

  if (process.env.NP_DEBUG) {
    fs.write('dmmf.json', dmmf)
  }

  const declarationSourceFile = ModuleGenerators.TS.createModule(dmmf, settings)

  if (settings.data.output.directory === 'default') {
    // ESM

    const esmSourceFiles = [
      ModuleGenerators.JS.createModule({
        gentimeSettings: settings,
        esm: true,
        dmmf,
      }),
      declarationSourceFile,
    ]

    // fs.remove(OUTPUT_SOURCE_DIR_ESM)

    esmSourceFiles.forEach((sf) => {
      const filePath = Path.join(OUTPUT_SOURCE_DIR_ESM, sf.fileName)
      fs.remove(filePath)
      fs.write(filePath, sf.content)
      d(`did write ${filePath}`)
    })

    // CJS

    const cjsSourceFiles = [
      ModuleGenerators.JS.createModule({
        gentimeSettings: settings,
        esm: false,
        dmmf,
      }),
      declarationSourceFile,
    ]

    fs.remove(OUTPUT_SOURCE_DIR_CJS)

    cjsSourceFiles.forEach((sf) => {
      const filePath = Path.join(OUTPUT_SOURCE_DIR_CJS, sf.fileName)
      fs.remove(filePath)
      fs.write(filePath, sf.content)
      d(`did write ${filePath}`)
    })
  } else {
    const sourceFiles = [
      ModuleGenerators.JS.createModule({
        gentimeSettings: settings,
        esm: false,
        dmmf,
      }),
      declarationSourceFile,
    ]

    // fs.remove(outputDir)

    sourceFiles.forEach((sf) => {
      const filePath = Path.join(
        settings.data.output.directory,
        sf.fileName.endsWith('d.ts') ? `${settings.data.output.name}.d.ts` : `${settings.data.output.name}.js`
      )
      fs.remove(filePath)
      fs.write(filePath, sf.content)
      d(`did write ${filePath}`)
    })
  }

  d(`done writing all emitted files`)
}

/** Transform the given DMMF into JS source code with accompanying TS declarations. */
export const generateRuntime = (dmmf: DMMF.Document, settings: Gentime.Settings.Manager): Module[] => {
  return [
    ModuleGenerators.JS.createModule({
      gentimeSettings: settings,
      esm: true,
      dmmf,
    }),
    ModuleGenerators.JS.createModule({
      gentimeSettings: settings,
      esm: false,
      dmmf,
    }),
    ModuleGenerators.TS.createModule(dmmf, settings),
  ]
}
