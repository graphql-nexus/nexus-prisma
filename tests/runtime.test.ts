import * as PrismaSDK from '@prisma/sdk'
import endent from 'endent'
import { generateRuntimeSource } from '../src/generator'

it('generates static TS code given DMMF', async () => {
  const schema = endent`
    datasource db {
      provider = "postgresql"
      url      = env("DB_URL")
    }

    model M1 {
      f10 String @id
      f11 Int
      f12 Float
      f13 Boolean
      f14 Decimal
      f15 BigInt
      f16 DateTime
      f17 Json
      f18 Bytes
    }
  `

  const dmmf = await PrismaSDK.getDMMF({
    datamodel: schema,
  })

  const source = generateRuntimeSource(dmmf)

  expect(source).toMatchSnapshot()
})
