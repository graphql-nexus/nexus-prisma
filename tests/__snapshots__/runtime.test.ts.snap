// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates static TS code given DMMF 1`] = `
"export const M1 = {
  f10: {
    name: 'f10',
    type: 'String',
    isList: false,
    isRequired: true,
    isId: true,
    isUnique: false,
    hasDefaultValue: false,
    documentation: null,
  },
  f11: {
    name: 'f11',
    type: 'Int',
    isList: false,
    isRequired: true,
    isId: false,
    isUnique: false,
    hasDefaultValue: false,
    documentation: null,
  },
  f12: {
    name: 'f12',
    type: 'Float',
    isList: false,
    isRequired: true,
    isId: false,
    isUnique: false,
    hasDefaultValue: false,
    documentation: null,
  },
  f13: {
    name: 'f13',
    type: 'Boolean',
    isList: false,
    isRequired: true,
    isId: false,
    isUnique: false,
    hasDefaultValue: false,
    documentation: null,
  },
  f14: {
    name: 'f14',
    type: 'Decimal',
    isList: false,
    isRequired: true,
    isId: false,
    isUnique: false,
    hasDefaultValue: false,
    documentation: null,
  },
  f15: {
    name: 'f15',
    type: 'BigInt',
    isList: false,
    isRequired: true,
    isId: false,
    isUnique: false,
    hasDefaultValue: false,
    documentation: null,
  },
  f16: {
    name: 'f16',
    type: 'DateTime',
    isList: false,
    isRequired: true,
    isId: false,
    isUnique: false,
    hasDefaultValue: false,
    documentation: null,
  },
  f17: {
    name: 'f17',
    type: 'Json',
    isList: false,
    isRequired: true,
    isId: false,
    isUnique: false,
    hasDefaultValue: false,
    documentation: null,
  },
  f18: {
    name: 'f18',
    type: 'Bytes',
    isList: false,
    isRequired: true,
    isId: false,
    isUnique: false,
    hasDefaultValue: false,
    documentation: null,
  },
} as const"
`;
